Add EV and empty capacitors

Prometheus0000 opened the pull request:
Empty has 0 storage, EV has the same as the IC2 lapotron crystal.

I'm thinking we can use the empty one like the AE2 storage housing for drives, while also allowing its use in the LSC as a filler block to reduce costs. The EV one can be used if you don't want to make the energy orb, and also reduces costs.

See https://github.com/GTNewHorizons/GT-New-Horizons-Modpack/issues/7090.

Prometheus0000 committed:
Add EV and empty capacitors

No recipes, reused textures for now

Prometheus0000 committed:
Add icons and recipes, and clean up spaces

Prometheus0000 commented:
![image](https://user-images.githubusercontent.com/8517039/115011272-7326fb80-9e7c-11eb-8abf-4e05eb4b2fef.png)
Empty, then EV-UHV. Clearly, I'm not an artist, so if anyone wants to make them fancier or w/e, that's fine.

empty
![image](https://user-images.githubusercontent.com/8517039/115011487-acf80200-9e7c-11eb-8b3d-e19ae0963c67.png)
ev
![image](https://user-images.githubusercontent.com/8517039/115011523-b71a0080-9e7c-11eb-81fe-1a829217fe65.png)
empty+crystal=ev
![image](https://user-images.githubusercontent.com/8517039/115011555-c39e5900-9e7c-11eb-9032-b69b4af3e089.png)
empty+orb=iv
![image](https://user-images.githubusercontent.com/8517039/115011569-c9943a00-9e7c-11eb-985e-0376fc0a0484.png)
Unpackaging gives the empty one for EV/IV instead of screws.
![image](https://user-images.githubusercontent.com/8517039/115011616-da44b000-9e7c-11eb-8054-2be8fe672f12.png)
![image](https://user-images.githubusercontent.com/8517039/115011645-e0d32780-9e7c-11eb-8786-f841ae112f4e.png)

Glease commented on src/main/java/common/tileentities/GTMTE_LapotronicSuperCapacitor.java:
```diff
@@ -233,14 +233,16 @@ public boolean checkMachine(IGregTechTileEntity thisController, ItemStack guiSlo
-							tempCapacity = tempCapacity.add(BigInteger.valueOf((long) (100000000L * Math.pow(10, 3))));
-							capacity = capacity.add(MAX_LONG);
-						}
+						switch(meta - 1) {
```
Why `- 1`?

Prometheus0000 replied:
Because of the ```final long c = (long) (100000000L * Math.pow(10, meta - 1));``` line. But I guess that's wrong, and I didn't notice it because the capacity is defined again correctly lower down.

Glease commented on src/main/java/common/tileentities/GTMTE_LapotronicSuperCapacitor.java:
```diff
@@ -314,14 +340,16 @@ public boolean checkMachine(IGregTechTileEntity thisController, ItemStack guiSlo
+			case 2: capacity = capacity.add(BigInteger.valueOf(10000000000L).multiply(BigInteger.valueOf(capacitors[i]))); break;
+			case 3: capacity = capacity.add(BigInteger.valueOf(100000000000L).multiply(BigInteger.valueOf(capacitors[i]))); break;
+			case 4: capacity = capacity.add(MAX_LONG.multiply(BigInteger.valueOf(capacitors[i]))); break;
+			case 5: capacity = capacity.add(BigInteger.valueOf(0L).multiply(BigInteger.valueOf(capacitors[i]))); break;
```
This line is no-op.

Prometheus0000 replied:
By no-op I guess you mean there doesn't need to be an operation?

bartimaeusnek replied:
``BigInteger.valueOf(0L).multiply(BigInteger.valueOf(capacitors[i]))``

=

``BigInteger.valueOf(0L) * capacitors.value()``

=

``BigInteger.valueOf(0L)``

and since this adds this value to the total capacity, you can just remove it. (default will catch it)

Prometheus0000 replied:
I realized that after looking at it, but I was unfamiliar with the terminology. Or do you mean removing the case entirely?

Glease reviewed (changes requested):
We probably need to define an minimal LSC capacity as well. Otherwise people will make LSCs made out of 100% empty cells (or with mostly empty cells) and call it an active transformer.

Prometheus0000 commented:
> 
> 
> We probably need to define an minimal LSC capacity as well. Otherwise people will make LSCs made out of 100% empty cells (or with mostly empty cells) and call it an active transformer.

Any suggestions on how to do this? I could check for X million EU, but 10 EV ones would be only 1 IV one. So it could be 180M, filling out a min size, but still only 2 of the IV ones.

bartimaeusnek commented on src/main/java/common/tileentities/GTMTE_LapotronicSuperCapacitor.java:
```diff
@@ -314,14 +340,16 @@ public boolean checkMachine(IGregTechTileEntity thisController, ItemStack guiSlo
-				capacity = capacity.add(
-						MAX_LONG.multiply(BigInteger.valueOf(capacitors[i])));
+		for(int i = 0; i < capacitors.length; i++){	
+			switch(i) {
```
You shouldn't use switches in functions. Wrap them please in a higher order function.

Prometheus0000 replied:
Why? It works fine?

Also, I don't understand what a HOF is, even after using google and reading some articles. Somehow it's different from a normal function? And there's a first one too? Also, every example used an array, which I don't really understand too well either, so I guess it needs that too? So I don't think I'd be able to do that.

bartimaeusnek commented on src/main/java/common/tileentities/GTMTE_LapotronicSuperCapacitor.java:
```diff
@@ -295,10 +297,34 @@ public boolean checkMachine(IGregTechTileEntity thisController, ItemStack guiSlo
-					formationChecklist = false;
-				}
-				break;
+				switch (highestCapacitor) {
```
You shouldn't use switches in functions. Wrap them please in a higher order function.

Prometheus0000 committed:
Fix

Prometheus0000 committed:
Remove the other one that doesn't do anything

Prometheus0000 committed:
change switch to function

Glease commented:
@Prometheus0000 the count of each kind of capacitor is stored in array `capacitors`. Maybe check if empty capacitor count is more than 50% (or something like that) of total capacitor used.

Prometheus0000 committed:
Revert previous and move the other 2 switches to functions

They actually work right, will redo the other one when I figure out how
to get it to do the structure check properly

Prometheus0000 committed:
Get 3rd switch to work in a function

Prometheus0000 committed:
Add check for more than 50% non-empty caps

Prometheus0000 committed:
Update TT text to reflect changes

Prometheus0000 commented:
Moved switches to functions, and added a check for > 50% being not empty. This could be changed to a lower amount if people aren't happy with it.

I did get an intermittent problem where the structure check still counted as valid when it shouldn't sometimes after adding the check, but I couldn't verify what caused it. Also, I'm not sure why it doesn't crash if you have all empty capacitors since it should be doing 18/0 for a min sized LSC. Nor could I find a way to deal with this using google if it did crash.

Glease commented on src/main/java/common/tileentities/GTMTE_LapotronicSuperCapacitor.java:
```diff
@@ -309,27 +299,87 @@ public boolean checkMachine(IGregTechTileEntity thisController, ItemStack guiSlo
 		}
 
+		//Check if enough (more than 50%) non-empty caps
+		int emptyCheck = capacitors[5]/(capacitors[0] + capacitors[1] + capacitors[2] + capacitors[3] + capacitors[4] + capacitors[6]);
```
This is not correct. You need to cast them to floating point numbers.
```java
double emptyCheck = ((double)capacitors[5])/(double)(capacitors[0] + capacitors[1] + capacitors[2] + capacitors[3] + capacitors[4] + capacitors[6]);

Prometheus0000 replied:
Fixed.

Prometheus0000 committed:
Change int to double to properly use fractions
